src/error.rs

```diff
@@ -296,6 +301,36 @@ impl StdError for ValidateError {
     }
 }
 
+/// Tag was corrupted during session.
+#[derive(Debug)]
+#[non_exhaustive]
+pub struct TagMismatch {
+    /// Expected tag number
+    pub(crate) expect: u32,
+    /// Actual tag number, 0 if parse failed
+    pub(crate) actual: u32,
```

Comment:
Let's make this a `Result<u32, Vec<u8>>` instead so we can directly present the unparseable tag if present.